NeoML/src/Dnn/BaseLayer.cpp

```diff
@@ -756,18 +771,32 @@ void CBaseLayer::CheckOutputs() const
 
 void CBaseLayer::onOutputProcessed( int index )
 {
-	if( !GetDnn()->isReuseMemoryMode ) {
-		return;
-	}
-
+	NeoPresume( GetDnn()->isReuseMemoryMode );
 	NeoPresume( outputProcessedCount.Size() > index );
 	NeoPresume( outputProcessedCount[index] < outputs[index] );
 
 	CPtr<CDnnBlob> result = outputBlobs[index];
 	outputProcessedCount[index]++;
-	if( outputProcessedCount[index] == outputs[index] ) {
+	if( outputProcessedCount[index] == outputs[index] && ( blobsNeededForBackward & TOutputBlobs ) == 0 ) {
 		outputBlobs[index] = 0;
 	}
 }
 
+void CBaseLayer::freeUnusedBlobs( int usedBlobs )
```

Comment:
freeUnused gets used, this is difficult to understand...

Reply:
I've tried to fix this... Not sure if it has become any clearer though